shotgun_api3/shotgun.py

```diff
@@ -3539,6 +3526,137 @@ def _dict_to_extra_data(self, d, key_name="value"):
         """
         return dict([(k, {key_name: v}) for (k,v) in (d or {}).iteritems()])
 
+    def _upload_file_to_storage(self, path, storage_url):
+        """
+        Internal function to upload an entire file to the Cloud storage.
+        
+        :param str path: Full path to an existing non-empty file on disk to upload.
+        :param str storage_url: Target URL for the uploaded file. 
+        """
+        filename = os.path.basename(path)
+
+        fd = open(path, "rb")
+        try:
+            content_type = mimetypes.guess_type(filename)[0]
+            content_type = content_type or "application/octet-stream"
+            file_size = os.fstat(fd.fileno())[stat.ST_SIZE]
+            self._upload_data_to_storage(fd, content_type, file_size, storage_url )
+        finally:
+            fd.close()
+
+        LOG.debug("File uploaded to Cloud storage: %s", filename)
+
+    def _multipart_upload_file_to_storage(self, path, upload_info):
+        """
+        Internal function to upload a file to the Cloud storage in multiple parts.
+
+        :param str path: Full path to an existing non-empty file on disk to upload.
+        :param dict upload_info: Contains details received from the server, about the upload.
+        """
+
+        fd = open(path, "rb")
+        try:
+            content_type = mimetypes.guess_type(path)[0]
+            content_type = content_type or "application/octet-stream"
+            file_size = os.fstat(fd.fileno())[stat.ST_SIZE]
+            filename = os.path.basename(path)
+
+            etags = []
+            part_number = 1
+            bytes_read = 0
+            chunk_size = 20000000
+            while bytes_read < file_size:
+                data = fd.read(chunk_size)
+                bytes_read += len(data)
+                part_url = self._get_upload_part_link(upload_info, filename, part_number)
+                etags.append(self._upload_data_to_storage(data, content_type, len(data), part_url ))
+                part_number += 1
+
+            self._complete_multipart_upload(upload_info, filename, etags)
```

Comment:
what happens if it fails half way? Should we worry about that case? should we send a message to the server saying that things broke half way and to discard the upload? Or will the server simply discard any "dangling" pieces after a certain amount of idle time?

I think a piece of DX to consider (and point out in the code) here is that we want a user to get the same experience when a multi-part upload fails and when a normal upload fails. Is this currently the case? Would be great with some tests to assert this.

I am thinking about the specific case where clients have flaky connections (not uncommon) so one data exchange with the server may work fine but the next one fails due to external factors.